### PR TITLE
✨ Products: Is3705/product groups  (🗃️ DB change)

### PR DESCRIPTION
**Pull request by pcrespov**
## What do these changes do?

- Products have an associated group (see ``products.group_id``) 
- Every user belong to one/more product's group with access ``read=False, write=False, delete=False``, i.e. they belong to the product but cannot see any other members of the products  (i.e. belongs but cannot share)
- User is automatically assigned to a product upon registration.

NOTE: @odeimaiz this PR still does NOT change the API, i.e. the ``list_products`` returns product groups as "non-readable" organizations. I would like to introduce the change in the API in a separate PR.

###  🗃️ DB change
 - adds ``products.group_id`` column
 - Defaults to NULL
 - Lazy creation of ``product.group_id`` upon registration of the first user (after these changes)
 - Current users need to be assigned manually! check with @pcrespov @odeimaiz when released.

## Related issue/s

- Implementing #3705

## How to test

- create several products
- add a service to every product
- register a new user with the default product (i.e. priority=0)
- the user has by default 3 groups: everybody groups (1), a product group and primary group 
- change default product
- user CAN login but does not see any

## Checklist

- [x] ``make version-*``
- [x] ``make openapi.json``
- [x] ``cd packages/postgres-database``, ``make setup-commit``, ``sc-pg review -m "my changes"``
- [x] Unit tests for the changes exist
- [x] Runs in the swarm
- [x] Documentation reflects the changes
- [x] New module? Add your github username to [.github/CODEOWNERS](.github/CODEOWNERS)

### CODE REVIEW

File: packages/postgres-database/src/simcore_postgres_database/migration/versions/1447a4f5d72b_add_product_s_group_column.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+# revision identifiers, used by Alembic.
+revision = "1447a4f5d72b"
+down_revision = "fe19b4e93586"
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.add_column("products", sa.Column("group_id", sa.BigInteger(), nullable=True))
```
Comment (sanderegg) [inline packages/postgres-database/src/simcore_postgres_database/migration/versions/1447a4f5d72b_add_product_s_group_column.py:20]: I do think we have ```gid``` everywhere don't we? maybe that would be nice to keep it the same for consistency?

File: packages/postgres-database/src/simcore_postgres_database/migration/versions/1447a4f5d72b_add_product_s_group_column.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    op.create_unique_constraint(
+        "products_group_id_unique_constraint", "products", ["group_id"]
+    )
+    op.create_foreign_key(
+        "fk_products_group_id",
+        "products",
+        "groups",
+        ["group_id"],
```
Comment (sanderegg) [inline packages/postgres-database/src/simcore_postgres_database/migration/versions/1447a4f5d72b_add_product_s_group_column.py:28]: same

File: packages/postgres-database/src/simcore_postgres_database/migration/versions/1447a4f5d72b_add_product_s_group_column.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    # ### end Alembic commands ###
+
+
+def downgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.drop_constraint("fk_products_group_id", "products", type_="foreignkey")
+    op.drop_constraint(
+        "products_group_id_unique_constraint", "products", type_="unique"
+    )
+    op.drop_column("products", "group_id")
```
Comment (sanderegg) [inline packages/postgres-database/src/simcore_postgres_database/migration/versions/1447a4f5d72b_add_product_s_group_column.py:42]: same

File: packages/postgres-database/src/simcore_postgres_database/models/products.py [hunk 4]
```diff
@@ -181,5 +182,18 @@ class Forum(TypedDict, total=True):
         nullable=True,
         doc="Limits the number of studies a user may have open concurently (disabled if NULL)",
     ),
+    sa.Column(
+        "group_id",
```
Comment (sanderegg) [inline packages/postgres-database/src/simcore_postgres_database/models/products.py:186]: I would keep the same ```gid``` term all over for consistency

File: packages/postgres-database/src/simcore_postgres_database/utils_products.py [hunk 5]
```diff
@@ -30,3 +43,42 @@ async def get_default_product_name(conn: _DBConnection) -> str:
...
+) -> Optional[int]:
+    group_id = await connection.scalar(
+        sa.select([products.c.group_id]).where(products.c.name == product_name)
+    )
+    return None if group_id is None else int(group_id)
+
+
+async def get_or_create_product_group(
+    connection: _AiopgConnection, product_name: str
+) -> int:
```
Comment (sanderegg) [inline packages/postgres-database/src/simcore_postgres_database/utils_products.py:59]: don't we have a GroupID alias somewhere? or is this not available in this package?

File: packages/postgres-database/tests/test_groups.py [hunk 6]
```diff
@@ -21,233 +24,234 @@
...
-        )
-        assert relations_count == (users_count + users_count)
+    rory_group = await create_fake_group(
+        connection, name="Rory Storm and the Hurricanes"
+    )
+    quarrymen_group = await create_fake_group(connection, name="The Quarrymen")
+    await create_fake_user(connection, name="John", group=quarrymen_group)
+    await create_fake_user(connection, name="Paul", group=quarrymen_group)
+    await create_fake_user(connection, name="Georges", group=quarrymen_group)
+    pete = await create_fake_user(connection, name="Pete", group=quarrymen_group)
```
Comment (sanderegg) [inline packages/postgres-database/tests/test_groups.py:179]: hope you learned something about the beatles here...

File: services/web/server/src/simcore_service_webserver/login/handlers_registration.py [hunk 7]
```diff
@@ -130,7 +130,10 @@ async def register(request: web.Request):
     )
 
     # NOTE: PC->SAN: should this go here or when user is actually logged in?
-    await auto_add_user_to_groups(request.app, user["id"])
+    await auto_add_user_to_groups(app=request.app, user_id=user["id"])
+    await auto_add_user_to_product_group(
```
Comment (sanderegg) [inline services/web/server/src/simcore_service_webserver/login/handlers_registration.py:134]: what about the NOTE here? didn't you want to change it?

**Review by sanderegg — APPROVED** [review]
:+1:

**Review by odeimaiz — APPROVED** [review]
👍

File: services/web/server/src/simcore_service_webserver/login/handlers_registration.py (hunk 7 above)
Comment (pcrespov) [reply services/web/server/src/simcore_service_webserver/login/handlers_registration.py:134]: Will do in a separate PR:
- https://github.com/ITISFoundation/osparc-simcore/issues/3709

File: packages/postgres-database/tests/test_groups.py (hunk 6 above)
Comment (pcrespov) [reply packages/postgres-database/tests/test_groups.py:179]: :-D your tests were very educational :-)

File: packages/postgres-database/tests/products/test_utils_products.py [hunk 8]
```diff
@@ -4,19 +4,96 @@
...
+            assert product_group.description == f"{product_row.name} product group"
+
+            # idempotent
+            for _ in range(3):
+                assert (
+                    await get_or_create_product_group(
+                        conn, product_name=product_row.name
+                    )
+                    == product_group_id
+                )
```
Comment (GitHK) [inline packages/postgres-database/tests/products/test_utils_products.py:73]: Maybe also add a count via `groups.name` to ensure this is truly idempotent. Where you expect the total amount of 1 entries

**Review by GitHK — APPROVED** [review]
Looks nice.

File: packages/postgres-database/tests/products/test_utils_products.py (hunk 8 above)
Comment (pcrespov) [reply packages/postgres-database/tests/products/test_utils_products.py:73]: that is tested a few lines below

File: packages/postgres-database/src/simcore_postgres_database/migration/versions/1447a4f5d72b_add_product_s_group_column.py (hunk 3 above)
Comment (pcrespov) [reply packages/postgres-database/src/simcore_postgres_database/migration/versions/1447a4f5d72b_add_product_s_group_column.py:42]: see above

File: packages/postgres-database/src/simcore_postgres_database/migration/versions/1447a4f5d72b_add_product_s_group_column.py (hunk 2 above)
Comment (pcrespov) [reply packages/postgres-database/src/simcore_postgres_database/migration/versions/1447a4f5d72b_add_product_s_group_column.py:28]: see above

File: packages/postgres-database/src/simcore_postgres_database/migration/versions/1447a4f5d72b_add_product_s_group_column.py (hunk 1 above)
Comment (pcrespov) [reply packages/postgres-database/src/simcore_postgres_database/migration/versions/1447a4f5d72b_add_product_s_group_column.py:20]: I never though i would say this but ... I would rather do the opposite, i.e. convert all ``gid`` into ``group_id``. I had a really bad time trying to match the indentifiers when i implemented the tags. My proposal:

- groups.id -> should be group's identifer in groups
- group_to

**Comment by mrnicegyu11** [comment]
lgtm :) thanks

**Pull request merged**